rs/bitcoin/kyt/src/fetch.rs

```diff
@@ -0,0 +1,270 @@
+// - Transactions bigger than 2MiB are very rare, and we can't handle them.
+
+/// Initial buffer size is 4kB
+pub const INITIAL_BUFFER_SIZE: u32 = 4 * 1024;
+
+/// Retry buffer size is 400kB
+pub const RETRY_BUFFER_SIZE: u32 = 400 * 1024;
+
+pub enum FetchResult<T> {
+    RetryWithBiggerBuffer,
+    Error(GetTxError),
+    Fetched(FetchedTx<T>),
+}
+
+pub enum TryFetchResult<T, F> {
+    Pending,
+    HighLoad,
+    Error(GetTxError),
+    NotEnoughCycles,
+    Fetched(FetchedTx<T>),
+    ToFetch(F),
+}
+
+pub trait HasOutPoint {
+    fn txid(&self) -> Txid;
+    fn vout(&self) -> u32;
+}
+
+pub trait TransactionLike {
+    type Input;
+    type AddressError;
+
+    fn iter_inputs(&self) -> impl Iterator<Item = &Self::Input>;
+    fn output_address(&self, vout: u32) -> Result<Address, Self::AddressError>;
+}
+
+/// Trait that abstracts over state operations.
+pub trait FetchState<T> {
+    type FetchGuard;
+    fn new_fetch_guard(&self, txid: Txid) -> Result<Self::FetchGuard, FetchGuardError>;
+    fn get_fetch_status(&self, txid: Txid) -> Option<FetchTxStatus<T>>;
+    fn set_fetch_status(&self, txid: Txid, status: FetchTxStatus<T>);
+    fn set_fetched_address(&self, txid: Txid, index: usize, address: Address);
+}
+
+/// Trait that abstracts over system functions like fetching transaction, calcuating cycles, etc.
+pub trait FetchEnv<I: HasOutPoint, T: Clone + TransactionLike<Input = I>>
+where
+    <T as TransactionLike>::AddressError: Debug,
+    T: Debug,
+{
+    async fn get_tx(&self, txid: Txid, buffer_size: u32) -> Result<T, GetTxError>;
+    fn cycles_accept(&self, cycles: u128) -> u128;
+    fn cycles_available(&self) -> u128;
+
+    /// Try to fetch a transaction given its txid:
+    /// - If it is already available, return `Fetched`.
+    /// - If it is already pending, return `Pending`.
+    /// - If it is pending retry or not found, return a future that calls `fetch_tx`.
+    /// - Or return other conditions like `HighLoad` or `Error`.
+    fn try_fetch_tx<State: FetchState<T>>(
+        &self,
+        state: &State,
+        txid: Txid,
+    ) -> TryFetchResult<T, impl futures::Future<Output = Result<FetchResult<T>, ()>>> {
+        let buffer_size = match state.get_fetch_status(txid) {
+            None => INITIAL_BUFFER_SIZE,
+            Some(FetchTxStatus::PendingRetry { buffer_size, .. }) => buffer_size,
+            Some(FetchTxStatus::PendingOutcall { .. }) => return TryFetchResult::Pending,
+            Some(FetchTxStatus::Error(msg)) => return TryFetchResult::Error(msg),
+            Some(FetchTxStatus::Fetched(fetched)) => return TryFetchResult::Fetched(fetched),
+        };
+        let guard = match state.new_fetch_guard(txid) {
+            Ok(guard) => guard,
+            Err(_) => return TryFetchResult::HighLoad,
+        };
+        let cycle_cost = get_tx_cycle_cost(buffer_size);
+        if self.cycles_accept(cycle_cost) < cycle_cost {
+            TryFetchResult::NotEnoughCycles
```

Comment:
I think it is better not to conflate the two ideas of guards (reentrancy prevention) and cycle accounting. It is not like that we'll do anything related to cycles when guards are dropped. What do you think?

Reply:
>  It is not like that we'll do anything related to cycles when guards are dropped.

Yes this makes sense.

> I think it is better not to conflate the two ideas of guards (reentrancy prevention) and cycle accounting.

I would usually agree. Here since the guard and cycles accounting is only done in one place I think both approaches are fine. If they were needed in more places I would feel more strongly about it, since cycles accounting is easily forgotten. Let's leave it as is.